src/pg_copy.c

```diff
@@ -0,0 +1,574 @@
+        sep = '(';
+        foreach(cell, copyStatement->options)
+        {
+            DefElem* def = (DefElem *) lfirst(cell);     
+            appendStringInfo(buf, "%c%s '%s'", sep, def->defname, defGetString(def));
+            sep = ',';
+        }
+        appendStringInfoChar(buf, ')');        
+    }   
+    return buf->data;
+}
+
+static bool PgCopyPrepareTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+	PQputCopyEnd(conn, NULL);
+	return tmgr->Prepare(conn, relationName, shardId);
+}
+
+static bool PgCopyAbortTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+
+	if (status) 
+	{
+		tmgr->RollbackPrepared(conn, relationName, shardId);
+	} else {
+		PQputCopyEnd(conn, "Aborted because of failure on some shard");
+		tmgr->Rollback(conn);
+	}
+	PQfinish(conn);
+	return true;
+}
+
+static bool PgCopyEndTransaction(ShardId shardId, PGconn* conn, void* arg, bool status)
+{
+	char const* relationName = (char*)arg;
+	PgShardTransactionManager const* tmgr = &PgShardTransManagerImpl[PgShardCurrTransManager];
+	
+	if (status)
+	{
+		tmgr->CommitPrepared(conn, relationName, shardId);
+	} else {
+		PQputCopyEnd(conn, "Aborted because of failure on some shard");
+		tmgr->Rollback(conn);
+	}
+	PQfinish(conn);
+	return true;
+}
+	
+
+static void 
+InitializeShardConnections(CopyStmt *copyStatement,
+						   ShardConnections *shardConnections,
+                           ShardId shardId,
+                           PgShardTransactionManager const* transactionManager) 
+{
+	ListCell *taskPlacementCell = NULL;
+	List *finalizedPlacementList = LoadFinalizedShardPlacementList(shardId); 
+	int placementCount = list_length(finalizedPlacementList);
+	
+	shardConnections->conn = (PGconn**)palloc0(sizeof(PGconn*)*placementCount);
+	shardConnections->replicaCount = placementCount;
+	shardConnections->status = (bool*)palloc0(placementCount*sizeof(bool));
+	placementCount = 0;
+					
+	foreach(taskPlacementCell, finalizedPlacementList)
+	{
+		ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+		char *nodeName = taskPlacement->nodeName;
+		PGconn* conn = NULL;
+		conn = ConnectToNode(nodeName, taskPlacement->nodePort);
+		if (conn != NULL)										 
+		{
+			char const* copy = ConstructCopyStatement(copyStatement, shardId);
+			shardConnections->conn[placementCount++] = conn;
+			/*
+			 * New connection: start transaction with copy command on it.
+			 * Append shard id to table name.
+			 */
+			if (!transactionManager->Begin(conn) ||
+				!PgShardExecute(conn, PGRES_COPY_IN, copy))
+			{
+				elog(ERROR, "Failed to start '%s' on node %s:%d", copy, nodeName, taskPlacement->nodePort);
```

Comment:
Why would data be missing on the second node? Isn't the all data written to all the placements?

By marking the placement as failed it is no longer included in the finalized placement list of subsequent selects, thus the data on the node we could not connect to is no longer visible.